core: allow to activate new BindPaths without unit restart

bluca opened the pull request:
Allow to add new BindPaths/BindReadOnlyPaths at runtime (via either
DBUS or systemctl set-property) with a new helper that forks into
the unit's mount namespace.
Add a new integration test to cover this.

Useful for zero-downtime addition to services that are running inside
mount namespaces, especially when using RootImage/RootDirectory.

bluca commented:
Latest push has more info in the manpage, and changed the test unit to be of Type=notify to ensure there's no race during activation, which is what I think could have caused the timeout on bionic-ppc64.

bluca commented:
New test failed once in bionic-ppc64 and once in bionic-s390 - given the journal artefacts are not saved, I've added TEST_SHOW_JOURNAL=debug which will hopefully print the journal if it fails again

bluca commented:
The MARKER_ messages are in the journal:

```
[ 1269.194073] cat[33]: MARKER_FIXED
...
[ 1270.221396] sh[63]: MARKER_RUNTIME
```

But for some reason journalctl | grep fails:

```
[ 1270.296892] testsuite-54.sh[68]: + grep -q -F MARKER_FIXED
[ 1270.297845] testsuite-54.sh[67]: + journalctl -b -u testservice-54.service
[ 1270.302854] systemd[1]: Received SIGCHLD from PID 29 (bash).
[ 1270.302898] systemd[1]: Child 29 (bash) died (code=exited, status=1/FAILURE)
```

I've switched the test to check for statuses instead of grepping the journal, should be more reliable.

poettering commented:
hmm, how does this work precisely? i mean, how to you get access to both the host fs and the service fs to be able to bind mount stuff across?

the per-service namespacing logic so far is optimized so that we suppress it if there is no bind mount nor any other mounts in place that would require namespacing for a service This means a simple service with just an ExecStart= line will not get a namespace of its own, but as soon as you add one namespace-based sandbox option, or one bind mount or such it will.

now this patch seems to ignore that: i could add a bind mount to any service i like regardless if it actually uses namespacing, and we cannot retroactively add namespacing for a running service...

Also, isn't there a vulnerability here: if there's already service code running and you manipulate the mount table of it while having access also to the host fs, then people might trick you into mounting stuff in a way you don't want, since mount() follows symlinks and they'd be resolved in the host fs, no?

what i also don't like: the current namespace code is quite careful: it sorts the mounts to establish so that upper mounts are always established before inner mounts, it does this while step-by-step following symlinks, so that symlinks can lie on other file systems. This means the order in which mounts are applied actually depends on symlinks found and where they point. But if we allow to retroactively establish mounts of course they'd always be applied last... Which is OK in general but might be confusing if exposed the same way as the original options, because they will behave systematically different depending if applied at service startup or a posteriori.

So I am not sure how this can work? it appears there are so many special cases here that it's too narrow to support?

What's the precise usecase here? can you elaborate? what's so bad with restarting the service?

maybe we can find a different solution, i.e. provide a separate tool that allows establishing mounts in arbitrary mount namespaces. Maybe called "systemd-namespace bind foo.service /some-place-on/host /some-place-in/container", that is a client to PID 1 and does the heavy lifting client side. And maybe later gains commands for copying files in and out too or so, similar to systemd-dissect. and it would generate a clean error when fs namespace for server and host is actually the same...

poettering commented:
hmm, there's prior art for all of this btw, in "machinectl bind". Maybe that's how we should do it here, too?

"machinectl bind" makes sure to not be vulnerable to the symlink attack i mentioned like this: nspawn guarantees that there's a runtime dir accessible from both the container and the host where propagation is enabled. machined then first bind mounts things into that dir, then enters the container fully with a tiny parasite process and moves the mount from there to the final destination. Now the trick is that the intermediare propagation dir is read-only when accessed by the container, which means the symlink attack cannot be done on the initial bind mount where it matters. it could be done for the second bind mount, but given that the parasite fully lives in the container's mount environment it doesn't matter there, symlinks would be resolved like they always are in the container.

bluca commented:
@poettering thanks for the review!

Lots to unpack here - first of all, the use case is zero-downtime addition/update of resources. I have core services where a restart means severe interruptions that just cannot happen unless absolutely necessary, ie: a kernel update or a glibc update and so on. If I could divulge the exact details it would be immediately obvious - sadly, I am not allowed yet and it's way, way above my pay grade. Sorry, I realise this sucks...
In terms of what the servicing does, there's 2 types: one is additional shared resources - access to persistent storage and between services is locked down with multiple layers of defense, so visibility is extremely restricted by default. Adding a new bind mount adds a new access, which can be enabled by the orchestrator at runtime.
The second type is updating individual files from the immutable application image - think of dlopen'ed shared libraries that get an update because of security bugs and the like. Again, zero-downtime is achieved with an over-mount, a notification and a new dlopen. This will be achieved by adding the same mechanism for the MountImages option - I haven't pushed it yet as I wanted to discuss the concept of runtime mounts first, with a smaller diff, but I plan to add support for that too.

The point about this working only if there's a namespace is absolutely correct, and I am aware of that - what happens is that set-property returns an error in that case (actually the check for this was missing, just added it). I think that's fine - the documentation does say that namespacing needs to be kept in mind, and I'll add a specific reference to this, which is missing from the diff.

Regarding ordering, I think it's understandable that a runtime addition only appends, it can't inject in the "middle" of a hierarchy. IMHO it's again a matter of make sure it's documented, which I've done explicitly now.

Then, to answer your question about how to get access to both the host and the guest filesystems, the helper function forks in the guest's namespace _without_ changing the root. So when the mount is activated, the process' ```/``` is the host's. To get the destination, we then use the file descriptor that represents the guest's root that namespace_open() gives us and get the path from there (ie: /proc/self/root) and use it as a prefix for the destination.

In terms of an external helper program - I wouldn't mind adding that too as an alternative that also works, but I must say the set-property path is something I'm quite keen on, for two reasons: first of all, it already understands that changes can be done persistently, so it both updates the unit with a drop-in and does the runtime change. I need this behaviour, as these changes are not transient - if there's a crash, the expectation of the orchestrator is that it doesn't need to re-run the command. Having a single, atomic thing to run is very desireable.
Secondly, it fits with the general modelling: there's one way to add mounts, and that's with the specific directive. Whether it also happens at runtime is an implementation detail that the orchestrator does not want to know and should not have to know. It's just setting a property of a service. It of course knows that a subset of properties can be changed at runtime and a subset can't, and that's fine - it's part of the interface. Having a subset that can't be changed at runtime, a subset that can be changed, and another subset that can be changed but only via a completely separate interface and that needs yet another interface for persistence doesn't sound very nice to me :-)

Finally, the symlink chasing is of course a very good point - I was under the impression that the target path symlink chase would solve it by removing the relative dereferencing, normalizing and then appending to the root directory - I guess that's not the case then? I refer to the follow_symlink() function here: https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L961
If it's not the case, I can factor out the "airlock" function from src/machine to shared/ so that it can be used in both places.

bluca commented:
> Finally, the symlink chasing is of course a very good point - I was under the impression that the target path symlink chase would solve it by removing the relative dereferencing, normalizing and then appending to the root directory - I guess that's not the case then? I refer to the follow_symlink() function here: https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L961
> If it's not the case, I can factor out the "airlock" function from src/machine to shared/ so that it can be used in both places.

Something I've been thinking and I don't quite understand: the mounting is happening inside the target process namespace, so it's not visible from the host. Even if via a relative symlink that doesn't get normalized the mountpoint ends up being in the host's root directory, what effect does it have? It's not in the host's namespace, no?

TriMoon commented:
@bluca not to sound mean or anything but it sounds like you want to be able to "Live patch" a running system and reboot it without an actual reboot? (Reboot used as exaggeration here)
I'm on poettering's side of view...

bluca commented:
> @bluca not to sound mean or anything but it sounds like you want to be able to "Live patch" a running system and reboot it without an actual reboot? (Reboot used as exaggeration here)
> I'm on poettering's side of view...

Not an entire system, but a portable service running with isolation mechanisms - namespaces, etc. Support in the applications to detect changes and reload is already there, I need to make things appear, which is what this is about. From a high level point of view, the use case is the same as kernel live patching really - zero downtime servicing.

poettering commented:
> Finally, the symlink chasing is of course a very good point - I was under the impression that the target path symlink chase would solve it by removing the relative dereferencing, normalizing and then appending to the root directory - I guess that's not the case then? I refer to the follow_symlink() function here: https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L961
> If it's not the case, I can factor out the "airlock" function from src/machine to shared/ so that it can be used in both places.

Yes, it's how this works currently. However this is only safe because we know the namespace is empty when we do it, i.e. noone else has access yet and thus cannot fuck with us. We cannot race against anyone else.

The problem is that mount() follows symlinks on the place where to mount, hence if someone could replace where we are supposed to mount something at the very moment where we finished chasing symlinks and are now doing mount() then we are fucked.

I mean, we generally have to trust RootImage=/RootDirectory= to some level because we mount them and that requires trust on Linux. However, trusting the OS image to some level is one thing. Trusting the code executed on it is another. And we should be really careful there: establishing mounts while some other (unpriv) code might do symlink tricks with us is problematic.

I am very concerned about the fact that this mounts established with your patch after the service is up would yield diferent results than those established right at the beginning. Because at the beginning we not only reorder mounts as we follow symlinks, we also suppress mounts based on InaccessiblePaths= further and stuff. I don't think how we can reasonably implement that correctly to yield the same results if you establish some paths late and some early...

I see your desire to have this robust, i.e. that on next start the mounts are established, but I think you could achieve this too by just dropping in some files in /etc/systemd/system/myservice.d/ manually first followed by "systemctl bind myservice /host/path /service/path" or so. In that case it's very clear that the the effect is not necessarily the same as the commands are very different.

I am just concerned that doing this via "systemctl set-property" would create the impression that it wouldn't matter if you change this when the service is started or when it is already running, but it actually matters a lot.

bluca commented:
> > Finally, the symlink chasing is of course a very good point - I was under the impression that the target path symlink chase would solve it by removing the relative dereferencing, normalizing and then appending to the root directory - I guess that's not the case then? I refer to the follow_symlink() function here: https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L961
> > If it's not the case, I can factor out the "airlock" function from src/machine to shared/ so that it can be used in both places.
> 
> Yes, it's how this works currently. However this is only safe because we know the namespace is empty when we do it, i.e. noone else has access yet and thus cannot fuck with us. We cannot race against anyone else.
> 
> The problem is that mount() follows symlinks on the place where to mount, hence if someone could replace where we are supposed to mount something at the very moment where we finished chasing symlinks and are now doing mount() then we are fucked.
> 
> I mean, we generally have to trust RootImage=/RootDirectory= to some level because we mount them and that requires trust on Linux. However, trusting the OS image to some level is one thing. Trusting the code executed on it is another. And we should be really careful there: establishing mounts while some other (unpriv) code might do symlink tricks with us is problematic.

Ok - I have done some refactoring and now the implementation is using the same airlock mechanism that machinectl/bind uses, with /run/systemd/propagate and /run/systemd/incoming. If this is a recurring problem, maybe it would be worth for mount(2) to have some sort of MS_DO_NOT_CHASE_SYMLINKS flag?

> I am very concerned about the fact that this mounts established with your patch after the service is up would yield diferent results than those established right at the beginning. Because at the beginning we not only reorder mounts as we follow symlinks, we also suppress mounts based on InaccessiblePaths= further and stuff. I don't think how we can reasonably implement that correctly to yield the same results if you establish some paths late and some early...
> 
> I see your desire to have this robust, i.e. that on next start the mounts are established, but I think you could achieve this too by just dropping in some files in /etc/systemd/system/myservice.d/ manually first followed by "systemctl bind myservice /host/path /service/path" or so. In that case it's very clear that the the effect is not necessarily the same as the commands are very different.
> 
> I am just concerned that doing this via "systemctl set-property" would create the impression that it wouldn't matter if you change this when the service is started or when it is already running, but it actually matters a lot.

I have added logic and a test case to ensure that InaccessiblePaths is respected, and anything that crosses it is dropped.

I have also added additional wording to the documentation regarding the ordering difference. Yes, that could be an issue, but on the other hand looking at it from the runtime perspective, it would seem intuitive that performing a new action on an existing runtime means the action happens on top of anything that was done at setup time, but changes to the configuration follow usual rules.

The main issue I have with writing drop-ins manually, apart from not being that nice&tidy from the interface design point of view (which is of course subjective), is that I'd need to give direct, blank-cheque write access to the orchestrator that performs these configurations, and I really really _really_ don't want to do that - instead of having everything mediated by systemd/portabled + selinux + polkit + dbus rules, which is great as permissions can be fine-tuned to a very granular level.

bluca commented:
Fixed merge conflict, and fixed a corner case: if the root/root img is read-only, then we can't create the airlock directory for incoming mounts. So if r/o is detected, a tmpfs is added on /run. As far as I can see, another mount on /run from the user will take priority over it, so there shouldn't be anything strange happening.

bluca commented:
One last corner case fixed, found by tests, to account for systemd --user - will use XDG_USER_RUNTIME instead of /run for propagate directories.

bluca commented:
changed the implementation as requested via new PR https://github.com/systemd/systemd/pull/17228
